### PR TITLE
[WIP] Change electric cables to use networks

### PR DESCRIPTION
**Pull request by wgraham17**
**Work in progress. This is not a backwards compatible change.**

Running Tech Reborn 1.15 on Fabric, we saw issues with long (100+ block) power networks where power would "ping pong" along the network and never reach the target machines. This is a change to not treat each individual cable block as a separate Energy entity but instead to treat a set of connected cables as a single network unit.

This is a **breaking change**.  Cables of a different tier will no longer directly connect and the use of transformers is required. Insulated and bare cables on the same tier will connect.

This is my first Java/Minecraft modding experience so please let me know if I need to make changes to ensure it doesn't negatively impact servers or clients.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -52,7 +52,7 @@ repositories {
 	}
 }
 
-version = "3.4.0"
+version = "3.3.11"
```
Comment (Sturmlilie) [inline build.gradle]: Squash this with the first commit.

File: src/main/java/techreborn/blockentity/cable/CableBlockEntity.java [hunk 2]
```diff
@@ -35,38 +35,35 @@
...
 import reborncore.common.network.ClientBoundPackets;
 import reborncore.common.network.NetworkManager;
 import reborncore.common.powerSystem.PowerAcceptorBlockEntity;
 import reborncore.common.powerSystem.PowerSystem;
 import reborncore.common.util.StringUtils;
-import team.reborn.energy.Energy;
-import team.reborn.energy.EnergySide;
-import team.reborn.energy.EnergyStorage;
-import team.reborn.energy.EnergyTier;
+import team.reborn.energy.*;
```
Comment (Sturmlilie) [inline src/main/java/techreborn/blockentity/cable/CableBlockEntity.java]: Why the change to `*`?

File: src/main/java/techreborn/blockentity/cable/CableBlockEntity.java [hunk 3]
```diff
@@ -35,38 +35,35 @@
...
 import techreborn.init.TRBlockEntities;
 import techreborn.init.TRContent;
 
 import java.util.ArrayList;
-import java.util.Collections;
 import java.util.List;
 
 /**
- * Created by modmuss50 on 19/05/2017.
+ * Created by wgraham17 on 28/05/2020.
```
Comment (Sturmlilie) [inline src/main/java/techreborn/blockentity/cable/CableBlockEntity.java]: Maybe add a `Modified by ...` line instead?

File: src/main/java/techreborn/blockentity/cable/CableBlockEntity.java [hunk 4]
```diff
@@ -35,38 +35,35 @@
...
  */
 
 public class CableBlockEntity extends BlockEntity
-	implements Tickable, IListInfoProvider, IToolDrop, EnergyStorage {
-	
-	private double energy = 0;
+		implements Tickable, IListInfoProvider, IToolDrop {
```
Comment (Sturmlilie) [inline src/main/java/techreborn/blockentity/cable/CableBlockEntity.java:64]: stray tab

File: src/main/java/techreborn/blockentity/cable/CableBlockEntity.java [hunk 5]
```diff
@@ -35,38 +35,35 @@
...
+
+	private int electric_net_id = 0;
```
Comment (Sturmlilie) [inline src/main/java/techreborn/blockentity/cable/CableBlockEntity.java]: surrounding code uses camelCase

File: src/main/java/techreborn/blockentity/cable/CableBlockEntity.java [hunk 6]
```diff
@@ -78,51 +75,83 @@ public CableBlockEntity(TRContent.Cables type) {
...
+	public TRContent.Cables getCableType() {
 		if (cableType != null) {
 			return cableType;
 		}
 		if (world == null) {
 			return TRContent.Cables.COPPER;
 		}
 		Block block = world.getBlockState(pos).getBlock();
-		if(block instanceof CableBlock){
+		if (block instanceof CableBlock) {
```
Comment (Sturmlilie) [inline src/main/java/techreborn/blockentity/cable/CableBlockEntity.java:87]: unrelated whitespace change

File: src/main/java/techreborn/blockentity/cable/CableBlockEntity.java [hunk 7]
```diff
@@ -78,51 +75,83 @@ public CableBlockEntity(TRContent.Cables type) {
...
+	public CompoundTag toInitialChunkDataTag() {
+		return toTag(new CompoundTag());
+	}
+
+	@Override
+	public BlockEntityUpdateS2CPacket toUpdatePacket() {
+		CompoundTag nbtTag = new CompoundTag();
+		toTag(nbtTag);
+		return new BlockEntityUpdateS2CPacket(getPos(), 1, nbtTag);
+	}
```
Comment (Sturmlilie) [inline src/main/java/techreborn/blockentity/cable/CableBlockEntity.java]: Unrelated indentation changes

File: src/main/java/techreborn/blockentity/cable/CableBlockEntity.java [hunk 8]
```diff
@@ -131,46 +160,17 @@ public CompoundTag toTag(CompoundTag compound) {
...
-
-        acceptors.forEach(pair -> {
-	        Energy.of(this)
-		        .into(Energy.of(pair.getLeft()).side(pair.getRight().getOpposite()))
-		        .move();
-        });
 	}
 
-    // IListInfoProvider
+	// IListInfoProvider
```
Comment (Sturmlilie) [inline src/main/java/techreborn/blockentity/cable/CableBlockEntity.java:174]: unrelated whitespace

File: src/main/java/techreborn/blocks/cable/CableBlock.java [hunk 9]
```diff
@@ -237,11 +241,10 @@ public void onEntityCollision(BlockState state, World worldIn, BlockPos pos, Ent
 			return;
 		}
 
-		CableBlockEntity blockEntityCable = (CableBlockEntity) blockEntity;
-		if (blockEntityCable.getEnergy() <= 0) {
+		if (!((CableBlockEntity)blockEntity).isEnergized()) {
 			return;
 		}
-
+
```
Comment (Sturmlilie) [inline src/main/java/techreborn/blocks/cable/CableBlock.java:247]: unrelated whitespace

File: src/main/java/techreborn/enet/PowerAcceptorBlockEntityFace.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
+package techreborn.enet;
```
Comment (Sturmlilie) [inline src/main/java/techreborn/enet/PowerAcceptorBlockEntityFace.java:25]: Missing license header?

**Review by Sturmlilie — COMMENTED** [review]
I'd suggest you to squash all your fix / cleanup commits.  
I'm not familiar with the energy code of TR so I couldn't comment on those specifics, but from a cursory look the logic appears sound. One thing that I'm a bit sad about is that (from my understanding), energy isn't distributed equally but on a first-serve basis, so the last consumers in an under-powered net get choked out. But that's a question of game design.

File: src/main/java/techreborn/blockentity/cable/CableBlockEntity.java [hunk 11]
```diff
@@ -189,74 +191,64 @@ public ItemStack getToolDrop(PlayerEntity playerIn) {
...
-		return energyOut;
 	}
 
-	public boolean canAcceptEnergy(EnergySide direction) {
-		if (sendingFace.contains(direction)) {
-			return false;
-		}
-		return getMaxStoredPower() != getEnergy();
+	public boolean canConnectTo(CableBlockEntity other) {
+		return other != null && other.getCableType().tier == getCableType().tier;
```
Comment (drcrazy) [inline src/main/java/techreborn/blockentity/cable/CableBlockEntity.java:206]: I thought that different cable types should be able to connect to each other.

File: src/main/java/techreborn/blocks/cable/CableBlock.java [hunk 12]
```diff
@@ -136,6 +136,10 @@ private BlockState makeConnections(World world, BlockPos pos) {
 	private Boolean canConnectTo(IWorld world, BlockPos pos, Direction facing) {
 		BlockEntity blockEntity = world.getBlockEntity(pos);
 		if (blockEntity != null && (Energy.valid(blockEntity) || blockEntity instanceof CableBlockEntity)) {
+			if (blockEntity instanceof CableBlockEntity && ((CableBlockEntity)blockEntity).getCableType().tier != type.tier) {
```
Comment (drcrazy) [inline src/main/java/techreborn/blocks/cable/CableBlock.java:139]: I thought that different cable types should be able to connect to each others.

File: src/main/java/techreborn/enet/ElectricNetwork.java [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+			TechReborn.LOGGER.debug(
+					"Visiting node {} on behalf of {} ({} visited total)",
+					currentEntity.getPos().asLong(),
+					this,
+					visitedNodes.size());
+
+			seenNodes.add(currentEntity);
+
+			for (Direction face : Direction.values()) {
+				BlockPos blockToCheckPos = currentEntity.getPos().offset(face);
```
Comment (drcrazy) [inline src/main/java/techreborn/enet/ElectricNetwork.java:160]: Did you have a chance to test how this will work when part of network is in unloaded chunk?

**Review by drcrazy — COMMENTED** [review]
Looks good. 
Ability to connect different tiers might be just a design limitation to ease transfer
And, please, double check chunk load\unload cases

File: build.gradle (hunk 1 above)
Comment (modmuss50) [reply build.gradle]: No need to squish commits in a PR, will be done when mergeing.

The version number should not be changed in the PR like this however.

File: src/main/java/techreborn/blockentity/cable/CableBlockEntity.java [hunk 14]
```diff
@@ -189,74 +191,64 @@ public ItemStack getToolDrop(PlayerEntity playerIn) {
...
 	}
 
-	@Override
-	public double getMaxInput(EnergySide side) {
-		if(!canAcceptEnergy(side)) {
-			return 0;
+	public ArrayList<PowerAcceptorBlockEntityFace> getConnectedPowerAcceptors() {
```
Comment (modmuss50) [inline src/main/java/techreborn/blockentity/cable/CableBlockEntity.java]: Set the return type to `List` here.

File: src/main/java/techreborn/blockentity/cable/CableBlockEntity.java [hunk 15]
```diff
@@ -189,74 +191,64 @@ public ItemStack getToolDrop(PlayerEntity playerIn) {
...
-	public double getMaxOutput(EnergySide side) {
-		return getCableType().transferRate;
+		return result;
 	}
 
-	@Override
-	public double getMaxStoredPower() {
-		return getCableType().transferRate * 4;
-	}
+	private void AssignNetwork() {
```
Comment (modmuss50) [inline src/main/java/techreborn/blockentity/cable/CableBlockEntity.java]: `assignNetwork`

File: src/main/java/techreborn/blocks/cable/CableBlock.java [hunk 16]
```diff
@@ -136,6 +136,10 @@ private BlockState makeConnections(World world, BlockPos pos) {
 	private Boolean canConnectTo(IWorld world, BlockPos pos, Direction facing) {
```
Comment (modmuss50) [inline src/main/java/techreborn/blocks/cable/CableBlock.java:136]: This method should really use the priamative type, not sure why i didnt.

I know you didnt add it in your PR but its an easy change to make a good time to do it.

File: src/main/java/techreborn/blocks/cable/CableBlock.java (hunk 12 above)
Comment (modmuss50) [reply src/main/java/techreborn/blocks/cable/CableBlock.java:139]: I quite like this idea, better than explosions, but there needs to be some form of backwards compatibliity.

File: src/main/java/techreborn/enet/ElectricNetwork.java [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+import techreborn.blockentity.cable.CableBlockEntity;
+
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public final class ElectricNetwork {
+	private final EnergyTier networkTier;
+	private final HashSet<CableBlockEntity> cableBlockEntities = new HashSet<>();
```
Comment (modmuss50) [inline src/main/java/techreborn/enet/ElectricNetwork.java]: Set the field type to `Set`

File: src/main/java/techreborn/enet/ElectricNetwork.java [hunk 18]
```diff
@@ -0,0 +1,238 @@
...
+
+		TechReborn.LOGGER.debug(
+				"Starting purge of disconnected entities. Have {}, seen {}",
+				cableBlockEntities.size(),
+				seenNodes.size());
+
+		cableBlockEntities
+				.stream()
+				.filter(node -> !seenNodes.contains(node))
+				.collect(Collectors.toList())
```
Comment (modmuss50) [inline src/main/java/techreborn/enet/ElectricNetwork.java:132]: Why this collect call?

File: src/main/java/techreborn/enet/ElectricNetwork.java [hunk 19]
```diff
@@ -0,0 +1,238 @@
...
+				.forEach(node -> {
+					if (node != null) {
```
Comment (modmuss50) [inline src/main/java/techreborn/enet/ElectricNetwork.java]: use filter to check for null node's

File: src/main/java/techreborn/enet/ElectricNetwork.java [hunk 20]
```diff
@@ -0,0 +1,238 @@
...
+								this);
+
+						node.setElectricNetwork(null);
+					}
+				});
+
+		TechReborn.LOGGER.debug("Finished walking network for cable connection changes.");
+	}
+
+	private HashSet<CableBlockEntity> visitCableNode(CableBlockEntity currentEntity, HashSet<CableBlockEntity> visitedNodes) {
```
Comment (modmuss50) [inline src/main/java/techreborn/enet/ElectricNetwork.java]: Again just set the return type and parmeters to `Set`, `HashSet` should still be used when creating the set.

File: src/main/java/techreborn/enet/ElectricNetwork.java (hunk 13 above)
Comment (modmuss50) [reply src/main/java/techreborn/enet/ElectricNetwork.java:160]: Yes, you need to be very careful here as it if a cable is on the edge of a chunk boarder its going to start loading the adjacent chunks causing all sorts of issues.

There is a way to check if a pos is in a loaded chunk but, id guess there is some more work to do around this.

File: src/main/java/techreborn/enet/ElectricNetwork.java [hunk 21]
```diff
@@ -0,0 +1,238 @@
...
+				.anyMatch(entityFace -> entityFace.canProvideEnergy());
+
+		transferPower(fromList, toList, networkTransferCapacity);
+	}
+
+	private void transferPower(List<PowerAcceptorBlockEntityFace> fromList, List<PowerAcceptorBlockEntityFace> toList, double networkTransferCapacity) {
+		double remainingTransfer = networkTransferCapacity;
+
+		for (PowerAcceptorBlockEntityFace to : toList) {
+			for (PowerAcceptorBlockEntityFace from : fromList) {
```
Comment (modmuss50) [inline src/main/java/techreborn/enet/ElectricNetwork.java:216]: This will need testing with a large amount of power acceptors and recievers to ensure that its not a burden on peformace.

File: src/main/java/techreborn/enet/ElectricNetworkManager.java [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+	public static void init() {
+		if (INSTANCE == null) {
+			INSTANCE = new ElectricNetworkManager();
+		}
+	}
+
+	@Override
+	public void tick(MinecraftServer minecraftServer) {
+		electricNetworks.addAll(pendingNetworks);
+		pendingNetworks.clear();
```
Comment (modmuss50) [inline src/main/java/techreborn/enet/ElectricNetworkManager.java]: Whats with `pendingNetworks`? why can the not be added right into `electricNetworks`?

File: src/main/java/techreborn/enet/ElectricNetworkManager.java [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+
+	public static void init() {
+		if (INSTANCE == null) {
+			INSTANCE = new ElectricNetworkManager();
+		}
+	}
+
+	@Override
+	public void tick(MinecraftServer minecraftServer) {
+		electricNetworks.addAll(pendingNetworks);
```
Comment (modmuss50) [inline src/main/java/techreborn/enet/ElectricNetworkManager.java]: I would make use of the vanilla profiler here to make it easy to see if its having an impact on tick peformace.

**Review by modmuss50 — CHANGES_REQUESTED** [review]
Overall looks like a solid start, there are a few style changes that should be made. Mainly using the base class of the Set/List/Map.

I like the idea of not allowing cables of diffrent tiers to connect to eatch other but there needs to be some sort of backwards compatiblity config, or setting in the world.

Great work 👍

File: src/main/java/techreborn/blockentity/cable/CableBlockEntity.java (hunk 3 above)
Comment (wgraham17) [reply src/main/java/techreborn/blockentity/cable/CableBlockEntity.java]: Unintended change; initially I had pulled these changes into a separate block entity altogether.

File: src/main/java/techreborn/enet/ElectricNetwork.java (hunk 18 above)
Comment (wgraham17) [reply src/main/java/techreborn/enet/ElectricNetwork.java:132]: This is needed because the `cableBlockEntities` collection can change when walking the network, so we need to materialize the results to avoid a concurrent modification exception.

**Comment by jusvit** [comment]
Closing due to inactivity, if you want to continue, there's a "feature/energynetwork" branch based off this PR with more work towards it for 1.16 which you can fork and expand on and then resubmit again as a PR.
